### ISSUE TITLE
Connection Timeout leads to data corruption

### ISSUE THREAD

**Issue by Dav1dde** [redis-rs/redis-rs#1252]
A connection is not invalidated when the connection timeout is reached. This is especially problematic when combined with a connection pool.

Scenario:

- Get connection from pool
- Set connection timeout 
- Execute `GET key1`
- Request times out, connection is returned back to the pool
- Server produces response for `GET key1`
- Get the same connection from the pool
- Execute `GET key2`
- Client returns value from `GET key1` request

Currently the connection is only invalidated on EOF: https://github.com/redis-rs/redis-rs/blob/255f7c2eb5f4cad6a295f8d5b1f9a9f1aafb2fe1/redis/src/connection.rs#L1244-L1274

I am not quite sure in which circumstances the connection can be kept open, but to be safe I think the best option is to always close the connection on an `IoError`.

Simple example (enable `disable-client-setinfo` feature):

```rs
use std::{
    io::{BufRead, BufReader, Write},
    net::TcpListener,
    thread,
    time::Duration,
};

use anyhow::Result;
use redis::{Commands, ConnectionLike, Value};

fn read_command(mut reader: impl BufRead) -> Result<(String, String)> {
    let mut discard = String::new();
    reader.read_line(&mut discard)?;
    reader.read_line(&mut discard)?;

    let mut command = String::new();
    reader.read_line(&mut command)?;
    command.truncate(command.len() - 2);

    reader.read_line(&mut discard)?;

    let mut arg = String::new();
    reader.read_line(&mut arg)?;
    arg.truncate(arg.len() - 2);

    Ok((command, arg))
}

fn fake_redis() -> Result<()> {
    let listener = TcpListener::bind("127.0.0.1:7777")?;

    for stream in listener.incoming() {
        let mut stream = stream?;
        println!("Got connection from {:?}", stream.peer_addr()?);

        let mut reader = BufReader::new(stream.try_clone()?);
        let (command, arg) = read_command(&mut reader)?;

        match (&*command, &*arg) {
            ("GET", "key1") => {
                thread::sleep(Duration::from_secs(1));
                stream.write_all(b"$4\r\nkey1\r\n")?;
            }
            ("GET", "key2") => {
                stream.write_all(b"$4\r\nkey2\r\n")?;
            }
            _ => println!("Invalid command {command} {arg}"),
        }
    }

    Ok(())
}

fn main() -> Result<()> {
    std::thread::spawn(fake_redis);

    let client = redis::Client::open("redis://127.0.0.1:7777/")?;
    let mut con = client.get_connection()?;
    con.set_read_timeout(Some(Duration::from_secs(1)))?;

    let _ = con.get::<_, Value>("key1");
    assert!(con.is_open());
    let value = con.get::<_, String>("key2")?;
    assert_eq!(value, "key2");

    Ok(())
}
```

Example fails with:

```
Got connection from 127.0.0.1:64977
thread 'main' panicked at src/main.rs:66:5:
assertion `left == right` failed
  left: "key1"
  right: "key2"
```

**Comment by Dav1dde** [comment]
I guess this is similar/duplicate to https://github.com/redis-rs/redis-rs/issues/325 but the resolution doesn't seem to apply here, since this is not using an `aio::Connection`.

**Comment by nihohit** [comment]
fixed in 0.27.4
